source/opt/local_ssa_elim_pass.h

```diff
@@ -0,0 +1,254 @@
+  // a load
+  bool IsLiveVar(uint32_t varId) const;
+
+  // Add stores using |ptr_id| to |insts|
+  void AddStores(uint32_t ptr_id, std::queue<ir::Instruction*>* insts);
+
+  // Delete |inst| and iterate DCE on all its operands. Won't delete
+  // labels. 
+  void DCEInst(ir::Instruction* inst);
+
+  // Return true if all uses of |varId| are only through supported reference
+  // operations ie. loads and store. Also cache in supported_ref_vars_;
+  bool HasOnlySupportedRefs(uint32_t varId);
+
+  // Return true if all uses of |id| are only name or decorate ops.
+  bool HasOnlyNamesAndDecorates(uint32_t id) const;
+
+  // Kill all name and decorate ops using |inst|
+  void KillNamesAndDecorates(ir::Instruction* inst);
+
+  // Kill all name and decorate ops using |id|
+  void KillNamesAndDecorates(uint32_t id);
+
+  // Initialize data structures used by LocalSSAElim for function |func|,
+  // specifically block predecessors and target variables.
+  void InitSSARewrite(ir::Function& func);
+
+  // Returns the id of the merge block declared by a merge instruction in 
+  // this block, if any.  If none, returns zero.
+  uint32_t MergeBlockIdIfAny(const ir::BasicBlock& blk);
+
+  // Compute structured successors for function |func|.
+  // A block's structured successors are the blocks it branches to
+  // together with its declared merge block if it has one.
+  // When order matters, the merge block always appears first.
+  // This assures correct depth first search in the presence of early 
+  // returns and kills. If the successor vector contain duplicates
+  // if the merge block, they are safely ignored by DFS.
+  void ComputeStructuredSuccessors(ir::Function* func);
+
+  // Compute structured block order for |func| into |structuredOrder|. This
+  // order has the property that dominators come before all blocks they
+  // dominate and merge blocks come after all blocks that are in the control
+  // constructs of their header.
+  void ComputeStructuredOrder(ir::Function* func,
+      std::list<ir::BasicBlock*>* order);
+
+  // Return true if loop header block
+  bool IsLoopHeader(ir::BasicBlock* block_ptr) const;
+
+  // Initialize SSA map for block with single predecessor
+  void SSABlockInitSinglePred(ir::BasicBlock* block_ptr);
+
+  // Return true if variable is loaded after the label
+  bool IsLiveAfter(uint32_t var_id, uint32_t label) const;
+
+  // Initialize SSA map for loop header block by merging SSA maps 
+  // from all predecessors. If any value ids differ for any variable
+  // across predecessors, create a phi function and use that value id
+  // for the variable in the new SSA map. Assumes all predecessors have
+  // been visited by SSARewrite except the back edge. Use a dummy value
+  // in the phi for the back edge until the back edge block is visited
+  // and patch the phi value then.
+  void SSABlockInitLoopHeader(std::list<ir::BasicBlock*>::iterator block_itr);
+
```

Comment:
Yeah, the poorly named SSABlockInitSelectionMerge routine actually handles all multi-predecessor non-loop-header blocks including continue blocks. I will fix the name and comment.